ThunkLibs/libX11/libX11_Guest.cpp

```diff
@@ -35,119 +38,198 @@ extern "C" {
 
 #include <vector>
 
+namespace {
+  constexpr static std::array<std::string_view, 13> CallbackKeys = {{
+    XNGeometryCallback,
+    XNDestroyCallback,
+    XNPreeditStartCallback,
+    XNPreeditDoneCallback,
+    XNPreeditDrawCallback,
+    XNPreeditCaretCallback,
+    XNPreeditStateNotifyCallback,
+    XNStatusStartCallback,
+    XNStatusDoneCallback,
+    XNStatusDrawCallback,
+    XNR6PreeditCallback,
+    XNStringConversionCallback,
+  }};
+
+  template<typename CallbackType>
+  static void ConvertArguments(std::vector<void*> &IncomingArguments, std::list<CallbackType> *Callbacks) {
+    // Walk the arguments and convert any callbacks.
+    const size_t ArgumentPairs = IncomingArguments.size() / 2;
+    for (size_t i = 0; i < ArgumentPairs; ++i) {
+      const char *Key = static_cast<const char*>(IncomingArguments[i * 2]);
+      void** Data = &IncomingArguments[i * 2 + 1];
+
+      // Check if the key is a callback and needs to be modified.
+      for (auto CallbackKey : CallbackKeys) {
+        if (CallbackKey == Key && *Data) {
```

Comment:
This can be flattened a lot:

```suggestion
      void** Data = &IncomingArguments[i * 2 + 1];
      if (!*Data) {
      	continue;
      }

      // Check if the key is a callback and needs to be modified.
      auto KeyIt = std::find(CallbackKeys.begin(), CallbackKeys.end(), Key);
      if (KeyIt == CallbackKeys.end()) {
      	continue;
      }
      
      // ...
```

Reply:
Flattened.